### DISCUSSION TITLE
How are resource files generated from Resources.l12n.xml

### DISCUSSION THREAD

**Discussion by HakanL** [unoplatform/uno#10470] [Q&A] [answered]
In the Uno ToDo sample project, there is a file called Resources.l12n.xml that has all the strings. Then there are 3 Resources.resw files in subfolders. How are these files generated from the xml file? Mine don't seem to re-generate when I modify the xml file, is there a custom tool or something I'm missing?

**Comment by jeromelaban** [comment]
@Xiaoy312 may be able to help.

**Comment by Xiaoy312** [comment, marked as answer]
This file was initially used to generated the other culture-specific xaml.
The associated script/tooling was not included as it was blocked by some unresolved internal discussion (we sort of forgot about it now...).

You can find the implementation here: // the PR commit history is messed up... just look at this single commit
https://github.com/unoplatform/uno.todo/pull/255/commits/1e788ffb0e6901dd53a2aefe55d2f4b5c41a6101
In the same PR description, I wrote an overview of it. I will just copy here:

<details>
<summary>Overview</summary>

Added `ReswGenerationTask` that triggers automatically on build of any head projects. This newly added task will regenerate .resw files based from the single master `resources.l12n.xml` file. This new localization format has a few advantages over the standard .resw file:
- Localization strings are stored in one file, and are grouped together:
  ```xml
  <data name="ImportantTaskListLabel">
    <en>Important</en>
    <fr>Important</fr>
    <es>Importante</es>
  </data>
  ```
- Localization strings can be grouped into `section`, and sections can nest themselves:
  > This makes navigation within the document easier, as most editors support xml folding.
  ```xml
  <section prefix="Dialog_">
    <data name="Common_Cancel">
      <en>Cancel</en>
      <fr>Annuler</fr>
      <es>Cancelar</es>
    </data>
    <section prefix="ConfirmDeleteList_">
      <data name="Title">
        <en>Are you sure?</en>
        <fr>En êtes-vous certain?</fr>
        <es>¿Estás seguro?</es>
      </data>
    </section>
  </section>
  ```
  which will generate into the following res: // only pick "en" for example
  ```xml
  <data name="Dialog_Common_Cancel" xml:space="preserve">
    <value>Cancel</value>
  </data>
  <data name="Dialog_ConfirmDeleteList_Title" xml:space="preserve">
    <value>Are you sure?</value>
  </data>
  ```
</details>

Of note, I'm not sure what state are the resw and l12n in now. The l12n may not have been kept updated. So just keep an eye.
Also, you can pull the target script out and run it separate. In fact, it was originally a linqpad script: https://github.com/Xiaoy312/bag-o-scripts/blob/main/temp/20220429%20l12n%20generator.linq

Hope that answers your question.

**Reply by Xiaoy312** [reply]
While we are here, there is another script for validating that the localization is done and done properly:
https://github.com/Xiaoy312/bag-o-scripts/blob/main/temp/20220429%20xaml%20l12n.linq
![image](https://user-images.githubusercontent.com/2359550/203853400-2bd800a9-3377-47a8-b749-2a4dd0dd4916.png)

You can adjust what should be localized in [the `GetLocalizableProperties` method](https://github.com/Xiaoy312/bag-o-scripts/blob/main/temp/20220429%20xaml%20l12n.linq#L77).
